File: Leetcode100/leetcode100_python/TrappingRainWater.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        result = 0
        n = len(height)
        max_left = [0 for _ in range(n)]
        max_right = [0 for _ in range(n)]
        for i in range(n):
            if i == 0:
                max_left[i] = height[i]
            else:
                max_left[i] = max(max_left[i - 1], height[i])
        for i in range(n - 1, -1, -1):
            if i == n - 1:
                max_right[i] = height[i]
            else:
                max_right[i] = max(max_right[i + 1], height[i])
        for i in range(n):
            result += min(max_left[i], max_right[i]) - height[i]
        return result

if __name__ == "__main__":
    height = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.trap(height))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/trapping_rain_water/TrappingRainWater.java
package main.java.com.jiang.leetcode100.trapping_rain_water;

import java.util.Scanner;
import java.util.Stack;

class Solution {
    public int trap(int[] height) {
        int result = 0;
        for(int i = 0; i < height.length; i++) {
            int heightest_left = height[i];
            for(int j = i; j < height.length; j++) {
                heightest_left = Math.max(heightest_left, height[j]);
            }

            int heightest_right = height[i];
            for(int j = i; j >= 0; j--) {
                heightest_right = Math.max(heightest_right, height[j]);
            }
            
            result += Math.min(heightest_left, heightest_right) - height[i];
        }

        return result;
    }
}

public class TrappingRainWater {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] height = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            height[i] = Integer.parseInt(nums_str[i]);
        }

        System.out.println(new Solution().trap(height));
    }
}